server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/FastVectorHighlighter.java

```diff
@@ -160,8 +160,13 @@ public HighlightField highlight(FieldHighlightContext fieldContext) throws IOExc
         }
         cache.fvh.setPhraseLimit(field.fieldOptions().phraseLimit());
 
-        String[] fragments;
+        // If the fragments builder requires document _source, pass it the source lookup from the hit context.
+        if (entry.fragmentsBuilder instanceof SourceBasedFragmentsBuilder) {
```

Comment:
This is hacky but felt like the best way to fix this right now. `FragmentsBuilder` is an external (Lucene) interface and there's not a good way to pass in _source to its methods like `getFields`.

Reply:
Could we replace the cached FragmentsBuilder with a `Function<SourceLookup, FragmentsBuilder>` and then call that with the new SourceLookup for each hit?  The FragmentsBuilders themselves don't seem to do any significant processing on construction so I don't think there would be a performance penalty for rebuilding them each time.

Reply:
It seems like that should work! I'd like to check I understand the context. We probably have these cached entries not because fragment builders are expensive to build, but to avoid rebuilding the `FieldQuery` which looks more complex?